Allow split actions to be edited and implement IOU action completeSplitBill

youssef-lr opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/Expensify/issues/322303


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

#### From Global Create

1. Start a split bill from Global Create by using the Scan tab and attaching a receipt, choose a new user.
2. Tap on the split action to reveal the split details page.
3. Make sure the page is editable.
4. Enter the amount, date, and merchant, make sure the amount is updated next to the 'Split' button.
5. Click 'Split', and confirm a report preview displaying the split amount.
6. Make sure an IOU report has been created, which should contain an expense with the same receipt attached to the original split action.

#### From Teachers Unite workspace

1. Sign up as a teacher using https://staging.new.expensify.com/teachersunite
2. Split a bill from the workspace chat created, and complete using the same test steps as above.
3. Make sure an expense report is created with half the amount of the total created.

#### Testing errors

1. Start a split bill
2. One the split details page, click on Split without entering the details
3. Make sure errors appear next to the missing field
4. Fill in those missing fields one by one, make sure the errors disappear, as well as the form error.
5. Complete the split bill and make sure child reports are created succesfully.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

The same test steps from above, but go offline before completing a split bill, then back online once you click on 'Split'.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
6. Upload an image via copy paste
7. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

This was tested thoroughly and evidence can be seen in the comments of this PR as well as in internal Slack channels where videos have been shared

<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

youssef-lr committed:
Make SplitBillDetailsPage editable when receipt is scanning or SS failed

youssef-lr committed:
Only show SmartScan fields splitting manually or editting a split

youssef-lr committed:
Create EditSplitBillPage

youssef-lr committed:
Create routes for EditSplitBillPage

youssef-lr committed:
Configure inputs on split details page and link them with routes

youssef-lr committed:
Save split transaction edits to a draft transaction

youssef-lr committed:
Cleanup

youssef-lr committed:
Add useCallback dep

youssef-lr committed:
Add proptype

youssef-lr committed:
Fix code removed by mistake

youssef-lr committed:
Bug fix and lint

youssef-lr committed:
wip

youssef-lr committed:
Merge branch 'youssef_startSplitBill' into youssef_completeSplitBill

mountiny commented on src/libs/TransactionUtils.ts:
```diff
@@ -96,7 +96,7 @@ function areRequiredFieldsEmpty(transaction: Transaction): boolean {
  * Given the edit made to the money request, return an updated transaction object.
  */
-function getUpdatedTransaction(transaction: Transaction, transactionChanges: TransactionChanges, isFromExpenseReport: boolean): Transaction {
+function getUpdatedTransaction(transaction: Transaction, transactionChanges: TransactionChanges, isFromExpenseReport: boolean, shouldUpdateReceiptState: boolean = true): Transaction {
```
Lint failing here
```suggestion
function getUpdatedTransaction(transaction: Transaction, transactionChanges: TransactionChanges, isFromExpenseReport: boolean, shouldUpdateReceiptState = true): Transaction {
```

mountiny reviewed (commented):
@youssef-lr lint is not happy

youssef-lr committed:
Merge branch 'youssef_startSplitBill' into youssef_completeSplitBill

youssef-lr committed:
Remove deprecated method and use withOnyx

youssef-lr committed:
Add completeSplitBill action

youssef-lr committed:
Send necessary params to completeSplitBill

youssef-lr committed:
Show right icon next to all fields in split details

youssef-lr committed:
Fix a few bugs

youssef-lr committed:
Merge branch 'main' into youssef_completeSplitBill

youssef-lr committed:
Cleanup

youssef-lr committed:
Update src/libs/TransactionUtils.ts

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

youssef-lr commented on src/libs/TransactionUtils.ts:
```diff
@@ -143,7 +143,13 @@ function getUpdatedTransaction(transaction: Transaction, transactionChanges: Tra
 
-    if (shouldStopSmartscan && transaction?.receipt && Object.keys(transaction.receipt).length > 0 && transaction?.receipt?.state !== CONST.IOU.RECEIPT_STATE.OPEN) {
+    if (
+        shouldUpdateReceiptState &&
```
I'm not sure why this function handles updating a receipt, the function name is `getUpdatedTransaction`, but seems like the function has a side effect.

youssef-lr committed:
Add nagivation to editting currency route

youssef-lr committed:
Add success & failure data to original transaction

youssef-lr committed:
Merge branch 'youssef_completeSplitBill' of github.com:Expensify/App into youssef_completeSplitBill

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+    const reportAction = ReportActionsUtils.getReportAction(report.reportID, lodashGet(route, ['params', 'reportActionID'], ''));
+    const transaction = TransactionUtils.getLinkedTransaction(reportAction);
+
+    let draftSplitTransaction = draftSplitTransactions[`${ONYXKEYS.COLLECTION.DRAFT_SPLIT_TRANSACTION}${transaction.transactionID}`];
```
Use `withOnyx`

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+    const transaction = TransactionUtils.getLinkedTransaction(reportAction);
+
+    let draftSplitTransaction = draftSplitTransactions[`${ONYXKEYS.COLLECTION.DRAFT_SPLIT_TRANSACTION}${transaction.transactionID}`];
+    if (!draftSplitTransaction) {
+        IOU.setDraftSplitTransaction(transaction.transactionID);
+    }
```
```suggestion
```

I don't think it's necessary to init the draft. The draft will be created once we make a call with `transactionChanges`

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+function EditSplitBillPage({report, route, draftSplitTransactions}) {
+    const fieldToEdit = lodashGet(route, ['params', 'field'], '');
+    const reportAction = ReportActionsUtils.getReportAction(report.reportID, lodashGet(route, ['params', 'reportActionID'], ''));
+    const transaction = TransactionUtils.getLinkedTransaction(reportAction);
```
After doing so the reportAction may no longer be needed. We only need the reportActionID

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+            <EditRequestDescriptionPage
+                defaultDescription={transactionDescription}
+                onSubmit={(transactionChanges) => {
+                    if (transactionChanges.comment === transactionDescription) {
```
```suggestion
                    if (transactionChanges.comment.trim() === transactionDescription) {
```

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+                        return;
+                    }
+                    setDraftSplitTransaction({
+                        comment: transactionChanges.comment,
```
```suggestion
                        comment: transactionChanges.comment.trim(),
```

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+                    const amount = CurrencyUtils.convertToBackendAmount(Number.parseFloat(transactionChanges));
+
+                    if (amount === transactionAmount && transactionCurrency === defaultCurrency) {
+                        Navigation.goBack();
```
```suggestion
                        Navigation.dismissModal();
```

I think you meant dismissModal, right?

youssef-lr replied:
No, we need to go back to the split details page, I'll change this to take the user back specifically to the route of the details in case of going back, in case they enter this page from a URL directly.

s77rt commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+            <EditRequestMerchantPage
+                defaultMerchant={transactionMerchant}
+                onSubmit={(transactionChanges) => {
+                    if (transactionMerchant === transactionChanges.merchant) {
+                        Navigation.goBack();
+                        return;
+                    }
+                    setDraftSplitTransaction({merchant: transactionChanges.merchant});
```
```suggestion
                    if (transactionMerchant === transactionChanges.merchant.trim()) {
                        Navigation.dismissModal();
                        return;
                    }
                    setDraftSplitTransaction({merchant: transactionChanges.merchant.trim()});
```

s77rt commented on src/pages/iou/SplitBillDetailsPage.js:
```diff
@@ -124,8 +156,17 @@ export default compose(
     withLocalize,
     withReportAndReportActionOrNotFound,
     withOnyx({
+        allTransactions: {
+            key: ONYXKEYS.COLLECTION.TRANSACTION,
+        },
+        draftSplitTransactions: {
+            key: ONYXKEYS.COLLECTION.DRAFT_SPLIT_TRANSACTION,
+        },
```
We are subscribing to the whole collections. We only need to subscribe to one key here. e.g: 

```js
        transaction: {
            key: ({route, reportActions}) => {
                const reportAction = reportActions[`${route.params.reportActionID.toString()}`];
                return `${ONYXKEYS.COLLECTION.TRANSACTION}${lodashGet(reportAction, 'originalMessage.IOUTransactionID', 0)}`;
            },
        },

```

youssef-lr replied:
Done

s77rt commented on src/pages/iou/SplitBillDetailsPage.js:
```diff
@@ -45,17 +50,25 @@ const propTypes = {
     }).isRequired,
 
+    /** Session info for the currently logged in user. */
+    session: PropTypes.shape({
+        /** Currently logged in user accountID */
+        accountID: PropTypes.number,
+    }).isRequired,
```
Add `email`

s77rt commented on src/pages/iou/SplitBillDetailsPage.js:
```diff
@@ -124,8 +156,17 @@ export default compose(
     withLocalize,
     withReportAndReportActionOrNotFound,
     withOnyx({
+        allTransactions: {
+            key: ONYXKEYS.COLLECTION.TRANSACTION,
+        },
```
Missing prop type

s77rt commented on src/pages/iou/SplitBillDetailsPage.js:
```diff
@@ -71,15 +84,30 @@ function SplitBillDetailsPage(props) {
+
+    const draftSplitTransaction = props.draftSplitTransactions && props.draftSplitTransactions[`${ONYXKEYS.COLLECTION.DRAFT_SPLIT_TRANSACTION}${transactionID}`];
+
+    if (isEdittingSplitBill && !draftSplitTransaction) {
+        IOU.setDraftSplitTransaction(transaction.transactionID);
+    }
```
```suggestion
```

No need to init with no changes

youssef-lr commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -202,13 +206,9 @@ function MoneyRequestConfirmationList(props) {
     // A flag and a toggler for showing the rest of the form fields
     const [shouldExpandFields, toggleShouldExpandFields] = useReducer((state) => !state, false);
-    const shouldShowAllFields = props.isDistanceRequest || shouldExpandFields || !shouldShowSmartScanFields;
+    const shouldShowAllFields = props.isDistanceRequest || shouldExpandFields || props.isEdittingSplitBill || !props.shouldShowSmartScanFields;
```
This component is also used when a split bill is being created and in this case it will be true even though we're not editing it.

youssef-lr committed:
Cleanup splitOrRequestOptions

youssef-lr commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -266,9 +266,9 @@ function MoneyRequestConfirmationList(props) {
     const splitOrRequestOptions = useMemo(() => {
         let text;
-        if (props.receiptPath && props.hasMultipleParticipants && props.iouAmount === 0) {
+        if (props.receiptPath && props.hasMultipleParticipants && props.isScanning && props.iouAmount === 0) {
```
Done, thanks!

youssef-lr committed:
Fix typo

youssef-lr committed:
Apply suggestions from code review

Co-authored-by: Abdelhafidh Belalia <16493223+s77rt@users.noreply.github.com>

youssef-lr commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -589,7 +616,7 @@ function MoneyRequestConfirmationList(props) {
                             description={translate('common.category')}
                             onPress={() => Navigation.navigate(ROUTES.MONEY_REQUEST_CATEGORY.getRoute(props.iouType, props.reportID))}
                             style={[styles.moneyRequestMenuItem]}
-                            disabled={didConfirm || props.isReadOnly}
+                            disabled={didConfirm}
```
`props.isReadOnly` is only set from the split details page, we decided we won't show these fields as disabled but instead remove the right icon and remove the navigation to the field. I need to update the rest.

youssef-lr replied:
I think I'll work on this in a follow up PR

youssef-lr committed:
Add failure data reverting changes made in optimistic data

youssef-lr committed:
Fix typo

youssef-lr commented on src/libs/actions/IOU.js:
```diff
@@ -1489,6 +1510,187 @@ function startSplitBill(participants, currentUserLogin, currentUserAccountID, co
+                },
+            },
+        },
+    ];
```
Done.

youssef-lr commented on src/libs/actions/IOU.js:
```diff
@@ -1489,6 +1510,187 @@ function startSplitBill(participants, currentUserLogin, currentUserAccountID, co
+            // Instead, we will rely on Auth generating the report IDs and the user won't see any optimistic chats or reports created
+            const participantPersonalDetails = allPersonalDetails[participant.accountID] || {};
+            if (!participantPersonalDetails || participantPersonalDetails.isOptimisticPersonalDetail) {
+                splits.push({
+                    email: participant.email,
+                });
```
It's not used in the backend yet, we'll need to refactor the backend down the line to use it. As for the split amount, it's also not used, we calculate it in the backend from the size of participants. It's used here to be displayed optimistically.

youssef-lr commented on src/libs/actions/IOU.js:
```diff
@@ -1489,6 +1510,187 @@ function startSplitBill(participants, currentUserLogin, currentUserAccountID, co
+            shouldCreateNewOneOnOneIOUReport,
+        );
+
+        splits.push({
+            email: participant.email,
+            accountID: participant.accountID,
```
Replied here https://github.com/Expensify/App/pull/29064#discussion_r1354143681

youssef-lr committed:
Optimize retrieval of Onyx data

youssef-lr committed:
Optimize Onyx data more

youssef-lr committed:
Address comments

youssef-lr commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,170 @@
+    /** Route from navigation */
+    route: PropTypes.shape({
+        /** Params from the route */
+        params: PropTypes.shape({
+            /** Which field we are editing */
+            field: PropTypes.string,
+
```
Done

youssef-lr committed:
Trim merchant

youssef-lr committed:
Always navigate back to split details

youssef-lr committed:
Fix jerky animation (might be reverted not sure if this is okay)

youssef-lr committed:
Show error fields next to missing fields

youssef-lr commented:
Updated @s77rt! All comments addressed and I just added displaying errors next to missing fields

https://github.com/Expensify/App/assets/9680864/596a1cfa-6df0-47ab-8350-e91c28039539

youssef-lr committed:
Linting

youssef-lr committed:
Remove testing code

youssef-lr committed:
Lint

youssef-lr committed:
Delete draft split transaction in success data

youssef-lr committed:
Display form error when SmartScan fails

youssef-lr committed:
Merge branch 'main' into youssef_completeSplitBill

youssef-lr committed:
Rename param

youssef-lr committed:
Include current user in splits

youssef-lr committed:
Lint

youssef-lr committed:
Dismiss modal and notify users after completing split bill

mountiny commented:
@s77rt you should be able to test against staging now

youssef-lr committed:
Update translations

mountiny commented on src/pages/EditRequestAmountPage.js:
```diff
@@ -35,17 +33,12 @@ function EditRequestAmountPage({defaultAmount, defaultCurrency, onSubmit, report
             }
-
-            textInput.current.focus();
+            setTimeout(() => {
```
Can you try using the same we have for the new request amount page which needs to handle the same transition coming from the Moneyrequest confirmation list? https://github.com/Expensify/App/blob/743c7b851b7acbf841c6e75f6814ac6005184276/src/pages/iou/steps/NewRequestAmountPage.js#L72-L84

youssef-lr replied:
ohh I missed this

mountiny commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,161 @@
+    /** The current transaction */
+    transaction: PropTypes.shape(transactionPropTypes).isRequired,
+
+    /** The draft transaction that holds data to be persisited on the current transaction */
```
```suggestion
    /** The draft transaction that holds data to be persisted on the current transaction */
```

mountiny commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,161 @@
+    route: PropTypes.shape({
+        /** Params from the route */
+        params: PropTypes.shape({
+            /** Which field we are editing */
```
```suggestion
            /** The transaction field we are editing */
```

youssef-lr committed:
Use existing timeout logic for focusing text input

youssef-lr committed:
Display form error when completing split bill with missing fields

youssef-lr committed:
Bug fix

youssef-lr committed:
Use areRequiredFieldsEmpty

youssef-lr committed:
Fix lint and change form error when user completes split bill

youssef-lr committed:
Display field errors next to missing smartscan fields when use confirms

youssef-lr committed:
Cleanup

youssef-lr commented on src/libs/actions/IOU.js:
```diff
@@ -1489,6 +1510,187 @@ function startSplitBill(participants, currentUserLogin, currentUserAccountID, co
+            created: updatedTransaction.modifiedCreated,
+            merchant: updatedTransaction.modifiedMerchant,
+            comment: updatedTransaction.comment.comment,
+            splits: JSON.stringify(splits),
```
If we don't stringify the object the backend ends up receiving `[Object object]` instead of the JSON object

youssef-lr committed:
Apply suggestions from code review

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

youssef-lr committed:
Revert change, should be handled in follow up clean up

youssef-lr commented:
Yeah working on that bug, I think I may need a backend change to fix it, although it's shown there it's actually not set in the transaction's modifedCreated, but only created.

youssef-lr commented:
Test steps updated.

s77rt commented:
Split requests are not editable when initialising 


<img width="378" alt="Screenshot 2023-10-11 at 2 45 21 PM" src="https://github.com/Expensify/App/assets/16493223/9df86cf2-2bd8-4783-a657-0d61646aa352">

youssef-lr commented:
That's expected @s77rt, we assume the user will rely on SmartScan to complete the split bill for them (which is implemented in the backend now).

s77rt commented:
Question: If the receipt is scanned, should we expect the options: amount, merchant, etc. to be disabled?

s77rt commented:
Minor bug: If we create the split request and click split quickly an error arises

https://github.com/Expensify/App/assets/16493223/29b68b21-7506-4ac2-af1a-6ef473024e19

youssef-lr commented:
Seems like an Onyx issue when the transaction is not ready yet in the component? will investigate

mountiny commented:
I will trigger a build now

s77rt commented:
> Minor bug: If we create the split request and click split quickly an error arises

After the receipt got scanned, getting same error and can't complete the split


https://github.com/Expensify/App/assets/16493223/84448d7c-d179-49a4-800c-6007a5f3f6bf

mountiny commented:
Isnt it because we only set the draft transaction when you edit something but if you just go to details and click split without changing anything then there is no draft yet which is where we grab the transaction from https://github.com/Expensify/App/blob/7a6f4b2d9c7c050867acc9cc3de115fd7afaa5b8/src/pages/EditSplitBillPage.js#L66-L69

youssef-lr commented:
If there's no draft transaction we still grab the original transaction though? investigating

youssef-lr commented:
ah just noticed that's in `completeSplitBill`, you're probably right @mountiny

OSBotify commented:
Use the links below to test this build in android and iOS. Happy testing!

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/29064/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/29064/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/29064/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/29064/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/29064/NewExpensify.dmg  | https://29064.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/29064/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://29064.pr-testing.expensify.com)  |

youssef-lr commented:
>After the receipt got scanned, getting same error and can't complete the split

How was that transaction got filled? you entered the details manually then clicked on 'split'?

s77rt commented:
It was filled by backend

youssef-lr commented:
Weird, when it gets filled by the backend it should no longer be editable

youssef-lr commented:
Are you testing against prod? There is a backend PR that just got deployed 16mins ago. Can you try again?

youssef-lr committed:
Fix isScanning state to also check for required fields being empty

youssef-lr commented:
Should be fixed now, can you try again please? Thanks for catching this!

youssef-lr commented:
Weird bug

<img width="388" alt="Screenshot 2023-10-11 at 22 54 57" src="https://github.com/Expensify/App/assets/9680864/82298457-0afe-452c-8739-3420064a9920">

youssef-lr committed:
Fix areRequiredFieldsEmpty to check for non existent modified fields

youssef-lr commented on src/libs/TransactionUtils.ts:
```diff
@@ -83,14 +83,19 @@ function hasReceipt(transaction: Transaction | undefined | null): boolean {
 }
 
 function areRequiredFieldsEmpty(transaction: Transaction): boolean {
```
cc @luacmartins  @mountiny can you guys check this change? `areRequiredFieldsEmpty` was returning false even tho we had a missing modifiedMerchant, because it doesn’t check for non existent modifiedMerchant in the object

s77rt commented:
> Should be fixed now, can you try again please? Thanks for catching this!

Once the receipt is scanned the options are filled automatically and disabled

youssef-lr commented:
Yup that's how it should work.

s77rt commented:
The first testing steps may be a little confusing

> Start a split bill from Global Create by using the Scan tab and attaching a receipt, choose a new user.
> Tap on the split action to reveal the split details page.
> Make sure the page is editable.

The first step is to start split bill and complete it (up until clicking Split)

s77rt commented:
On mWeb the split details is not scrollable


https://github.com/Expensify/App/assets/16493223/ce9c2098-f45c-48d8-92b9-ed642a743998

s77rt commented:
On mweb/native the form error overlaps with the rest of elements

<img width="422" alt="Screenshot 2023-10-11 at 5 26 19 PM" src="https://github.com/Expensify/App/assets/16493223/93e566b9-0605-4dd6-b4e2-9b2b60f0ed09">

s77rt commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/62170b25-c330-4568-a694-4a60ba92488b


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/07d2fd3b-c44a-4810-a7b1-eaed2c42e152


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/c0746c50-e846-464c-905c-9c2131aa7cb7


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/c9e69383-c905-4d6e-b648-b9c6af3b4d70


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/0e0fe0dc-527d-4b85-bd55-1ba1373bd55b


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/5527ec56-537a-4356-a24a-c44bd9f1b7e5


</details>

youssef-lr commented:
@s77rt any other issues you found other than the two UI issues on mWeb?

youssef-lr commented:
Can you check please if the issue exists with regular splits as well?

youssef-lr committed:
Use non-interactive style instead of disabled style

youssef-lr commented:
Weird, why is linting failed based on files I haven't changed in this PR?

<img width="1073" alt="Screenshot 2023-10-12 at 02 36 41" src="https://github.com/Expensify/App/assets/9680864/126477ac-4947-4137-948c-fd412e9b2830">

youssef-lr committed:
Merge branch 'main' into youssef_completeSplitBill

s77rt commented:
> any other issues you found other than the two UI issues on mWeb?

So far that's all.

> Can you check please if the issue exists with regular splits as well?

They don't (or at least I cannot reproduce them - maybe because of the much more available space)

s77rt commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -510,26 +548,45 @@ function MoneyRequestConfirmationList(props) {
                     titleStyle={styles.moneyRequestConfirmationAmount}
-                    disabled={didConfirm || props.isReadOnly}
+                    disabled={didConfirm}
+                    brickRoadIndicator={shouldDisplayFieldError && transaction.modifiedAmount === 0 && CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR}
+                    error={shouldDisplayFieldError && transaction.modifiedAmount === 0 && translate('common.error.enterAmount')}
```
This is causing a prop warning. `brickRoadIndicator` and `error` return booleans if the condition is not met

luacmartins commented:
@youssef-lr we just fixed the lint error in [this PR](https://github.com/Expensify/App/pull/29359). Please merge main again

s77rt commented on src/pages/iou/SplitBillDetailsPage.js:
```diff
@@ -34,6 +36,12 @@ const propTypes = {
     /** Array of report actions for this report */
     reportActions: PropTypes.shape(reportActionPropTypes),
 
+    /** The current transaction */
+    transaction: PropTypes.shape(transactionPropTypes).isRequired,
```
Getting prop warning that the transaction is undefined
<img width="564" alt="Screenshot 2023-10-11 at 8 23 56 PM" src="https://github.com/Expensify/App/assets/16493223/4dfca0bf-3d91-4b81-b6fd-1211df017662">

youssef-lr committed:
Merge branch 'main' into youssef_completeSplitBill

s77rt commented:
Not getting an error when the amount is unset

<img width="375" alt="Screenshot 2023-10-11 at 8 29 34 PM" src="https://github.com/Expensify/App/assets/16493223/30cca102-f38b-44a0-b490-e7d889b813ab">

youssef-lr committed:
Fix MenuItem console warning and checking for modified amount

youssef-lr commented:
@s77rt fixed the console warnings related to MenuItem as well as the amount issue, still trying to figure out the transaction prop warning issue.

youssef-lr commented:
TIL
<img width="326" alt="Screenshot 2023-10-12 at 04 13 08" src="https://github.com/Expensify/App/assets/9680864/7e50c770-cbd6-4cb2-bdab-36fb18475beb">

youssef-lr committed:
Fix checking for modifiedAmount

youssef-lr committed:
Keep merchant error if it equal partial marchant (none)

s77rt commented:
Wow really interesting! I expected to false too

youssef-lr commented:
I have no idea how to fix this..

<img width="792" alt="Screenshot 2023-10-12 at 04 24 14" src="https://github.com/Expensify/App/assets/9680864/1b68d97c-74c9-40fc-9a63-08495d120cc3">

maybe it's due to how we fetch the transaction from Onyx using reportActions that are also from Onyx?

<img width="1305" alt="Screenshot 2023-10-12 at 04 24 59" src="https://github.com/Expensify/App/assets/9680864/2d7e00a3-7ae7-443c-88f8-766a89f91f7a">

s77rt commented:
Strange how the error is `The prop transaction.isRequired is marked as required` instead of `The prop transaction is marked as required`.

s77rt commented:
Okay got it. It should be

```js
transaction: transactionPropTypes.isRequired
```

and not 

```js
transaction: PropTypes.shape(transactionPropTypes).isRequired
```


`transactionPropTypes` already is a shape. Same fix needed for `draftTransaction` in `SplitBillDetailsPage` and `EditSplitBillPage`

mountiny committed:
Merge branch 'main' into youssef_completeSplitBill

mountiny committed:
Fix console warning

mountiny commented:
@s77rt this is ready, can you pleaase complete the checklist so we can get this PR to staging today. Thanks!

s77rt commented:
Conflicts

s77rt commented:
@mountiny Can you please fix this prop type for draftTransaction as well

s77rt commented:
Is the mweb bug supposed to be fixed later?

<img width="564" alt="Screenshot 2023-10-12 at 1 03 08 AM" src="https://github.com/Expensify/App/assets/16493223/dabd953d-eb74-44bb-acb3-e1f93ead127f">

luacmartins committed:
resolve conflicts

luacmartins commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,161 @@
+        comment: transactionDescription,
+        merchant: transactionMerchant,
+        created: transactionCreated,
+    } = draftTransaction ? ReportUtils.getTransactionDetails(draftTransaction) : ReportUtils.getTransactionDetails(transaction);
```
This will always return true because we add a default `{}`, no?

youssef-lr replied:
Will remove the default as the draft shouldn't be required

youssef-lr replied:
Will make it undefined in the follow up PR

luacmartins commented:
Note: we merged this PR as is and will address the remaining issues as a followup tomorrow @youssef-lr

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

youssef-lr commented:
Thanks folks!!!

youssef-lr commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -182,17 +182,19 @@ const defaultProps = {
-    const {onSendMoney, onConfirm, onSelectParticipant, transaction} = props;
+    const {onSendMoney, onConfirm, onSelectParticipant} = props;
     const {translate, toLocaleDigit} = useLocalize();
+    const transaction = props.isEditingSplitBill ? props.draftTransaction || props.transaction : props.transaction;
```
Working on a follow up PR to clean things up, we wanted this merged quickly so we can test 🙏🏼

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6513406442) to production by https://github.com/jasperhuangg in version: 1.3.83-11 🚀

platform | result
---|---
🤖 android 🤖|skipped 🚫
🖥 desktop 🖥|skipped 🚫
🍎 iOS 🍎|skipped 🚫
🕸 web 🕸|skipped 🚫

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6513406442) to production by https://github.com/jasperhuangg in version: 1.3.83-11 🚀

platform | result
---|---
🤖 android 🤖|skipped 🚫
🖥 desktop 🖥|skipped 🚫
🍎 iOS 🍎|skipped 🚫
🕸 web 🕸|skipped 🚫

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6513406442) to production by https://github.com/jasperhuangg in version: 1.3.83-11 🚀

platform | result
---|---
🤖 android 🤖|skipped 🚫
🖥 desktop 🖥|skipped 🚫
🍎 iOS 🍎|skipped 🚫
🕸 web 🕸|skipped 🚫

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6513766880) to production by https://github.com/jasperhuangg in version: 1.3.83-11 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

DylanDylann commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -518,26 +556,45 @@ function MoneyRequestConfirmationList(props) {
-                    onPress={() => !props.isDistanceRequest && Navigation.navigate(ROUTES.MONEY_REQUEST_AMOUNT.getRoute(props.iouType, props.reportID))}
+                    interactive={!props.isReadOnly}
+                    onPress={() => {
+                        if (props.isDistanceRequest) {
+                            return;
+                        }
```
@youssef-lr I have a minor question: why do we avoid user edit amount in case of distance request? Please help answer once you have a chance.

tgolen commented on src/pages/EditSplitBillPage.js:
```diff
@@ -0,0 +1,161 @@
+        return (
+            <EditRequestCreatedPage
+                defaultCreated={transactionCreated}
+                defaultAmount={transactionAmount}
+                reportID={reportID}
```
FYI, this component doesn't use either of these props. I'm going to remove them in my PR and I'll ping you on it.

s77rt commented on src/components/MoneyRequestConfirmationList.js:
```diff
@@ -576,18 +642,36 @@ function MoneyRequestConfirmationList(props) {
+                                    ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR
+                                    : ''
+                            }
+                            error={
+                                shouldDisplayFieldError && (transaction.modifiedMerchant === '' || transaction.modifiedMerchant === CONST.TRANSACTION.PARTIAL_TRANSACTION_MERCHANT)
+                                    ? translate('common.error.enterMerchant')
+                                    : ''
+                            }
```
We should show the merchant error only if the merchant is required i.e. the report is linked to a workspace.
